consumer/consumererror/networkerrors.go

```diff
@@ -0,0 +1,89 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package consumererror // import "go.opentelemetry.io/collector/consumer/consumererror"
+
+import (
+	"fmt"
+	"net/http"
+
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	"go.opentelemetry.io/collector/consumer/consumererror/internal/statusconversion"
+)
+
+// NetworkError contains either an HTTP or gRPC status code
+// resulting from a failed synchronous network request.
+// It allows retrieving a type of error code, converting
+// between the status codes supported by each transport if necessary.
+//
+// It should be created with NewHTTPStatus or NewGRPCStatus.
+type NetworkError struct {
+	error
+	httpStatus *int
+	grpcStatus *status.Status
+}
+
+var _ error = NetworkError{}
+
+// Error returns a string specifying the transport and corresponding status code.
+func (se NetworkError) Error() string {
+	if se.httpStatus != nil {
+		return fmt.Sprintf("HTTP Status (%d): %s", *se.httpStatus, se.error.Error())
+	} else if se.grpcStatus != nil {
+		return fmt.Sprintf("gRPC Status (%s): %s", se.grpcStatus.Code().String(), se.error.Error())
+	}
+
+	return "Network error (no error code set): " + se.error.Error()
+}
+
+func (se NetworkError) Unwrap() error {
+	return se.error
+}
+
+// HTTPStatus returns an HTTP status code either directly
+// set by the source or derived from a gRPC status code set
+// by the source.
+// If no code has been set, the second return value is
+// an HTTP 500 code.
+func (se NetworkError) HTTPStatus() int {
+	if se.httpStatus != nil {
+		return *se.httpStatus
+	} else if se.grpcStatus != nil {
+		return statusconversion.GetHTTPStatusCodeFromStatus(se.grpcStatus)
+	}
+
+	return http.StatusInternalServerError
+}
+
+// GRPCStatus returns an gRPC status code either directly
+// set by the source or derived from an HTTP status code set
+// by the source.
+// If no code has been set, the second return value is set
+// to `false`.
+func (se NetworkError) GRPCStatus() *status.Status {
+	if se.grpcStatus != nil {
+		return se.grpcStatus
+	} else if se.httpStatus != nil {
+		return statusconversion.NewStatusFromMsgAndHTTPCode(se.error.Error(), *se.httpStatus)
+	}
+
+	return status.New(codes.Unknown, se.Error())
+}
+
+// NewHTTPError wraps an error with a given HTTP status code.
+func NewHTTPError(err error, code int) error {
```

Comment:
Should all of the new `New*` functions for the new error types take options so that we don't have to break the function signature in the future?

Reply:
I'm split on this. On one hand, that seems like a good idea. On the other, without a clear use-case in mind right now, I'm worried it will end up being dead code. Can you think of any potential options we could add to these errors?

Reply:
_maybe_ a `WithLogger` option for network errors if we ever wanted to add debug statements or something.  I don't really have a good idea in mind.

With our updated policy on variadic param additions are we covered if we don't do it now?

Reply:
>_maybe_ a `WithLogger` option for network errors if we ever wanted to add debug statements or something.  I don't really have a good idea in mind.

For most cases I could see a logger possibly being helpful, but I think we should avoid putting a logger inside our error structs. Thanks for brainstorming on it.

>With our updated policy on variadic param additions are we covered if we don't do it now?

The policy only allows us to avoid the deprecation process, but either way we're covered until 1.0.

Reply:
Based on @dmitryax's `source` idea, I'm back to thinking Options would be a good idea.

Reply:
I added `consumererror.Component` to cover this since we may also want this information for exporting retries and I think it makes more sense to have a top-level object instead of repeating the same option across two different error types.